packages/ai-core/src/chatTransport.ts

```diff
@@ -0,0 +1,282 @@
+  apiKey: string;
+  baseUrl?: string;
+  headers?: Record<string, string>;
+  getInstructions: () => string;
+  /**
+   * Optional: supply a pre-configured custom model.
+   * e.g. import {xai} from "@ai-sdk/xai";
+   * getCustomModel: () => xai('grok-4')
+   * If provided, this model will be used instead of the default OpenAI-compatible client.
+   */
+  getCustomModel?: () => LanguageModel | undefined;
+};
+
+export function createLocalChatTransportFactory({
+  get,
+  defaultProvider,
+  defaultModel,
+  apiKey,
+  baseUrl,
+  headers,
+  getInstructions,
+  getCustomModel,
+}: ChatTransportConfig) {
+  return () => {
+    const fetchImpl = async (_input: RequestInfo | URL, init?: RequestInit) => {
+      // Resolve provider/model and client at call time to pick up latest settings
+      const state = get();
+      const currentSession = state.ai.getCurrentSession();
+      const provider = currentSession?.modelProvider || defaultProvider;
+      const modelId = currentSession?.model || defaultModel;
+
+      // Prefer a user-supplied model if available
+      let model: LanguageModel | undefined = getCustomModel?.();
+
+      // Fallback to OpenAI-compatible if no custom model provided
+      if (!model) {
+        const openai = createOpenAICompatible({
+          apiKey,
+          name: provider,
+          baseURL: baseUrl || 'https://api.openai.com/v1',
+          headers,
+        });
+        model = openai.chatModel(modelId);
+      }
+
+      const body = init?.body as string;
+      const parsed = body ? JSON.parse(body) : {};
+      const messagesCopy = JSON.parse(JSON.stringify(parsed.messages || []));
+
+      // update the onToolCompleted handler to update the tool additional data in the store
+      const onToolCompleted = (toolCallId: string, additionalData: unknown) => {
+        const sessionId = get().ai.config.currentSessionId;
+        if (!sessionId) return;
+
+        get().ai.setSessionToolAdditionalData(
+          sessionId,
+          toolCallId,
+          additionalData,
+        );
+      };
+
+      const tools = Object.entries(state.ai.tools || {}).reduce(
+        (acc: ToolSet, [name, tool]: [string, OpenAssistantTool]) => {
+          acc[name] = convertToVercelAiToolV5({
+            ...tool,
+            onToolCompleted,
+          });
+          return acc;
+        },
+        {},
+      );
+
+      // get system instructions dynamically at request time to ensure fresh table schema
+      const systemInstructions = getInstructions();
+
+      const result = streamText({
+        model,
+        messages: convertToModelMessages(messagesCopy),
+        tools,
+        system: systemInstructions,
+        abortSignal: state.ai.analysisAbortController?.signal,
+      });
+
+      return result.toUIMessageStreamResponse();
+    };
+
+    return new DefaultChatTransport({fetch: fetchImpl});
+  };
+}
+
+export function createRemoteChatTransportFactory(params: {
+  get: GetAiSliceState;
```

Comment:
same as above, maybe pass store: `StoreApi<AiSliceState>` instead of `GetAiSliceState`